### PR TITLE
Reconcile definitions of stability.

### PR DESCRIPTION
**Pull request by paulp**

### CODE REVIEW

File: src/reflect/scala/reflect/internal/Types.scala [hunk 1]
```diff
@@ -2105,7 +2105,7 @@ trait Types extends api.Types { self: SymbolTable =>
   class ModuleTypeRef(pre0: Type, sym0: Symbol) extends NoArgsTypeRef(pre0, sym0) with ClassTypeRef {
     require(sym.isModuleClass, sym)
     private[this] var narrowedCache: Type = _
-    override def isStable = true
+    override def isStable = pre.isStable
```
Comment (xeno-by) [inline src/reflect/scala/reflect/internal/Types.scala:2108]: ?Was that a bug waiting to happen? Can we come up with a unit test for that?

Comment (retronym) [inline src/reflect/scala/reflect/internal/Types.scala:2108]: Such types show up after erasure.

``` scala
scala> :power
** Power User mode enabled - BEEP WHIR GYVE **
** :phase has been set to 'typer'.          **
** scala.tools.nsc._ has been imported      **
** global._, definitions._ also imported    **
** Try  :help, :vals, power.<tab>           **

scala> class C { object O }
defined class C

scala> typeOf[c.O.type forSome { val c: C }]
warning: there were 1 feature warning(s); re-run with -feature for details
res2: $r.intp.global.Type = c.O.type forSome { val c: C }

scala> res2.underlying
res3: $r.intp.global.Type = c.O.type

scala> .erasure
res4: $r.intp.global.Type = C#O.type

scala> res4.getClass
res5: Class[_ <: $r.intp.global.Type] = class scala.reflect.internal.Types$ModuleTypeRef
```

But that can't be what was guarded against in `Implicits`.

In any case:

``` patch
diff --git a/src/reflect/scala/reflect/internal/Types.scala b/src/reflect/scala/reflect/internal/Types.s
index fb493fa..f2982bc 100644
--- a/src/reflect/scala/reflect/internal/Types.scala
+++ b/src/reflect/scala/reflect/internal/Types.scala
@@ -2100,7 +2100,7 @@ trait Types extends api.Types { self: SymbolTable =>
   class ModuleTypeRef(pre0: Type, sym0: Symbol) extends NoArgsTypeRef(pre0, sym0) with ClassTypeRef {
     require(sym.isModuleClass, sym)
     private[this] var narrowedCache: Type = _
-    override def isStable = true
+    override def isStable = {assert(pre0.isStable, s"ModuleTypeRef#isStable, which is hard coded to ret
     override def narrow = {
       if (narrowedCache eq null)
         narrowedCache = singleType(pre, sym.sourceModule)
```

```
// Among others.
Log file '/Users/jason/code/scala2/test/files/pos/caseaccs-pos.log': 
Possible compiler crash during test of: /Users/jason/code/scala2/test/files/pos/caseaccs.scala
java.lang.AssertionError: assertion failed: ModuleTypeRef#isStable, which is hard coded to return `true`, called with unstable prefix: Test#Foo.type.
    at scala.Predef$.assert(Predef.scala:179)
    at scala.reflect.internal.Types$ModuleTypeRef.isStable(Types.scala:2103)
    at scala.tools.nsc.typechecker.Typers$Typer.stabilize(Typers.scala:651)
    at scala.tools.nsc.typechecker.Typers$Typer.scala$tools$nsc$typechecker$Typers$Typer$$typedSelect$1(Typers.scala:4863)
    at scala.tools.nsc.typechecker.Typers$Typer.typedSelectOrSuperCall$1(Typers.scala:4936)
    at scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:5549)
    at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:839)
```

Comment (retronym) [inline src/reflect/scala/reflect/internal/Types.scala:2108]: The change doesn't affect the generated bytecode of that test. During erasure, when `Foo.this.$outer` is typed, the qualifier `Foo.this` has type `Test#Foo.type` after the change, and `Foo.this.type` before.

If Paul could confirm that `scala-library.jar` is byte-for-byte unchanged after this, then lgtm.

Comment (paulp) [inline src/reflect/scala/reflect/internal/Types.scala:2108]: I'll check the bytecode. Even if such types don't presently show up before erasure, there are multiple possible encodings of the "X#Y.type" in class X { object Y } (see recent scala-internals discussion) and the most straightforward one would be TypeRef(X, Y, Nil), at which point you'd need this check.

Comment (paulp) [inline src/reflect/scala/reflect/internal/Types.scala:2108]: Verified identical bytecode.

**Pull request merged**